pkg/crypto/certloader/config.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package certloader
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+// WatchedConfig is a set of TLS configuration files including CA files, and a
+// certificate along with its private key. The files are watched for change and
+// reloaded automatically.
+type WatchedConfig struct {
+	log         logrus.FieldLogger
+	caFiles     []string
+	certFile    string
+	privkeyFile string
+	stop        chan struct{}
+	mu          lock.Mutex
+	caCertPool  *x509.CertPool
+	keypair     *tls.Certificate
+}
+
+var (
+	// ErrCertWithoutPrivkey is returned when a certificate is configured
+	// without its corresponding private key.
+	ErrCertWithoutPrivkey = errors.New("certificate and private key are both required, but only the certificate was provided")
+	// ErrPrivkeyWithoutCert is returned when a private key is configured
+	// without its corresponding certificate.
+	ErrPrivkeyWithoutCert = errors.New("certificate and private key are both required, but only the private key was provided")
+)
+
+// NewWatchedConfig returns a WatchedConfig that watch over the given file
+// paths. When caFiles is nil or empty, the system CA CertPool is used.
+func NewWatchedConfig(log logrus.FieldLogger, caFiles []string, certFile, privkeyFile string) (*WatchedConfig, error) {
+	if certFile != "" && privkeyFile == "" {
+		return nil, ErrCertWithoutPrivkey
+	}
+	if certFile == "" && privkeyFile != "" {
+		return nil, ErrPrivkeyWithoutCert
+	}
+
+	cfg := &WatchedConfig{
+		log:         log,
+		caFiles:     caFiles,
+		certFile:    certFile,
+		privkeyFile: privkeyFile,
+	}
+
+	// load the files for the first time before starting the watcher and update
+	// goroutine.
+	if err := cfg.Reload(); err != nil {
+		return nil, err
+	}
+	if err := cfg.Watch(); err != nil {
+		return nil, err
+	}
```

Comment:
If this is taken out of the constructor, a `WatchedConfig` could also be used in a more "manual" way where a caller does **not** call `Watch()` but calls `Reload()` upon receiving `SIGHUP` event for instance.
Essentially, the caller would have control over watching for cert changes or not.

Reply:
I think you have a point. On the other hand, the suggested approach open the door to forgetting a `.Watch()` call, so it is a trade-off.

I would prefer to take it when there is an actual use-case for "manual" use. For example, we could imagine an alternate constructor that doesn't call `Watch` to be used "manually", but it make little sense to implement it before there is a use-case for it.

Reply:
I've refactored the `certloader` module, and now the `Watcher` and `Reloader` functionalities are split. The use-case described by @Rolinh should be easy to implement using only the `Reloader`.